File: Study/programmers_42578.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

int solution(vector<vector<string>> clothes) {
    int answer = 1;
    unordered_map <string,int> map;
    vector<string> categories;
    for(int i = 0; i< clothes.size(); i++){
        string item = clothes[i][1];
        if(!map[item]){
            categories.push_back(item);
        }
        map[item]++;
    }

    for(string category : categories){
        answer *= map[category] + 1; // 1을 더하는 것은 해당 카테고리 의상을 입지 않는 경우를 고려하기 위해 
    }
    answer-= 1; // 해당 카테고리 의상을 입지 않는 경우를 고려하여 조합을 계산했으니 의상을 모두 미착용한 경우를 제거

    return answer;
}

File: 오답노트/programmers_42578.js
//BEST 풀이
function solution(clothes) {
  return (
    Object.values(
      clothes.reduce((obj, t) => {
        obj[t[1]] = obj[t[1]] ? obj[t[1]] + 1 : 1; //객체에 {카테고리: 옷 갯수} 형태로 갯수를 누적
        return obj;
      }, {})
    ).reduce((a, b) => a * (b + 1), 1) - 1 //그 갯수들을 + 1 한 값으로 모두 곱한 후 -1 (조합 수식)
  );
}
//나의풀이
function solution(clothes) {
  var answer = 1;
  const map = new Map();
  clothes.forEach((item) => {
    map.set(item[1], (map.get(item[1]) || 1) + 1); //입지 않는 경우의 수 1을 포함해서 count
  });
  //수학/about.md 의 조합 참고
  map.forEach((i) => {
    // 조합 구하는 공식
    answer *= i;
  });
  return answer - 1; // 아무것도 안입는 경우 제외
}
